File: packages/oxlint-plugin-react-doctor/src/plugin/rules/correctness/no-impure-call-at-module-scope.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { findVariableInitializer } from "../../utils/find-variable-initializer.js";
import { isFunctionLike } from "../../utils/is-function-like.js";
import { stripParenExpression } from "../../utils/strip-paren-expression.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";

// Per the applied revision, `crypto.*` id/byte generators are dropped:
// stable per-process ids are the dominant, correct idiom at module scope.
// What remains fires regardless of intent: Math.random() sampling and
// wall-clock reads used in date/timezone math.
const IMPURE_MEMBER_CALLS = new Map<string, ReadonlySet<string>>([
  ["Math", new Set(["random"])],
  ["Date", new Set(["now"])],
  ["performance", new Set(["now"])],
]);

const impureBuiltinLabel = (node: EsTreeNode): string | null => {
  if (isNodeOfType(node, "NewExpression")) {
    // Only the zero-argument `new Date()` is nondeterministic; a
    // timestamp/parts argument is deterministic.
    if (
      isNodeOfType(node.callee, "Identifier") &&
      node.callee.name === "Date" &&
      (node.arguments?.length ?? 0) === 0 &&
      !findVariableInitializer(node.callee, "Date")
    ) {
      return "new Date()";
    }
    return null;
  }
  if (!isNodeOfType(node, "CallExpression")) return null;
  if (
    isNodeOfType(node.callee, "Identifier") &&
    node.callee.name === "Date" &&
    node.arguments.length === 0 &&
    !findVariableInitializer(node.callee, "Date")
  ) {
    return "Date()";
  }
  const callee = stripParenExpression(node.callee);
  if (!isNodeOfType(callee, "MemberExpression") || callee.computed) return null;
  const receiver = stripParenExpression(callee.object);
  if (!isNodeOfType(receiver, "Identifier")) return null;
  if (!isNodeOfType(callee.property, "Identifier")) return null;
  const allowedMethods = IMPURE_MEMBER_CALLS.get(receiver.name);
  if (!allowedMethods?.has(callee.property.name)) return null;
  // A same-file binding named `Math`/`Date`/`performance` shadows the global.
  if (findVariableInitializer(receiver, receiver.name)) return null;
  return `${receiver.name}.${callee.property.name}()`;
};

const serverValueOfTypeofBrowserGlobalTest = (test: EsTreeNode): boolean | null => {
  const expression = stripParenExpression(test);
  if (!isNodeOfType(expression, "BinaryExpression")) return null;
  const isUndefinedTypeof = (candidate: EsTreeNode): boolean =>
    isNodeOfType(candidate, "UnaryExpression") &&
    candidate.operator === "typeof" &&
    isNodeOfType(candidate.argument, "Identifier") &&
    (candidate.argument.name === "window" || candidate.argument.name === "document");
  const isUndefinedLiteral = (candidate: EsTreeNode): boolean =>
    isNodeOfType(candidate, "Literal") && candidate.value === "undefined";
  if (
    !(
      (isUndefinedTypeof(expression.left as EsTreeNode) && isUndefinedLiteral(expression.right)) ||
      (isUndefinedTypeof(expression.right) && isUndefinedLiteral(expression.left as EsTreeNode))
    )
  ) {
    return null;
  }
  if (expression.operator === "===" || expression.operator === "==") return true;
  if (expression.operator === "!==" || expression.operator === "!=") return false;
  return null;
};

const isModuleScopeEvaluation = (impureNode: EsTreeNode, context: RuleContext): boolean => {
  let child: EsTreeNode = impureNode;
  let cursor: EsTreeNode | null = impureNode.parent ?? null;
  let isStaticField = false;
  while (cursor) {
    if (isFunctionLike(cursor) || isNodeOfType(cursor, "MethodDefinition")) return false;

    // `typeof window === "undefined" ? 0 : performance.now()` — the branch
    // visible to the server render is the deterministic constant; the
    // impure read only ever runs in the browser.
    if (isNodeOfType(cursor, "ConditionalExpression") && cursor.test !== child) {
      const serverTestValue = serverValueOfTypeofBrowserGlobalTest(cursor.test as EsTreeNode);
      if (
        serverTestValue !== null &&
        ((cursor.consequent === child && !serverTestValue) ||
          (cursor.alternate === child && serverTestValue))
      ) {
        return false;
      }
    }

    if (isNodeOfType(cursor, "PropertyDefinition")) {
      if (cursor.static !== true || cursor.key === child) return false;
      isStaticField = true;
    }

    if (isNodeOfType(cursor, "VariableDeclarator")) {
      const declaration = cursor.parent;
      if (!declaration || !isNodeOfType(declaration, "VariableDeclaration")) return false;
      if (declaration.kind !== "const") {
        const symbol = context.scopes.symbolFor(cursor.id);
        if (!symbol || symbol.references.some((reference) => reference.flag !== "read")) {
          return false;
        }
      }
      let declarationParent = declaration.parent ?? null;
      if (declarationParent && isNodeOfType(declarationParent, "ExportNamedDeclaration")) {
        declarationParent = declarationParent.parent ?? null;
      }
      if (!declarationParent || !isNodeOfType(declarationParent, "Program")) return false;
      return true;
    }

    child = cursor;
    cursor = cursor.parent ?? null;
  }
  return isStaticField;
};

export const noImpureCallAtModuleScope = defineRule({
  id: "no-impure-call-at-module-scope",
  title: "Nondeterministic built-in at module scope",
  severity: "warn",
  requires: ["ssr"],
  tags: ["test-noise"],
  recommendation:
    "`Math.random()`, `Date.now()`, `performance.now()`, and `new Date()` run once at module load, so the value is frozen for the whole server process. Move the call into a function/component so it evaluates per request.",
  create: (context: RuleContext) => {
    const check = (node: EsTreeNode): void => {
      const label = impureBuiltinLabel(node);
      if (!label) return;
      if (!isModuleScopeEvaluation(node, context)) return;
      context.report({
        node,
        message: `\`${label}\` runs once when this module loads, so the value is frozen for the whole server process and every SSR request reuses it — move it into a function or component so it evaluates per request.`,
      });
    };
    return {
      CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
        check(node);
      },
      NewExpression(node: EsTreeNodeOfType<"NewExpression">) {
        check(node);
      },
    };
  },
});

File: native/oxlint/rules/no-impure-call-at-module-scope.rs
use oxc_ast::{
    AstKind,
    ast::Expression,
};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;
use oxc_syntax::operator::{BinaryOperator, LogicalOperator, UnaryOperator};

use crate::{
    AstNode,
    context::{ContextHost, LintContext},
    rule::Rule,
};

const IMPURE_MEMBER_CALLS: [(&str, &str); 3] = [
    ("Math", "random"),
    ("Date", "now"),
    ("performance", "now"),
];

#[derive(Debug, Default, Clone)]
pub struct NoImpureCallAtModuleScope;

declare_oxc_lint!(
    /// Disallow nondeterministic built-ins evaluated at module scope.
    NoImpureCallAtModuleScope,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "Disallow nondeterministic module-scope built-ins.",
);

impl Rule for NoImpureCallAtModuleScope {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        !is_non_production_file(ctx)
    }

    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let Some(label) = impure_builtin_label(node, ctx) else {
            return;
        };
        if !is_module_scope_evaluation(node, ctx) {
            return;
        }
        ctx.diagnostic(
            OxcDiagnostic::warn(format!(
                "`{label}` runs once when this module loads, so the value is frozen for the whole server process and every SSR request reuses it — move it into a function or component so it evaluates per request."
            ))
            .with_label(node.span()),
        );
    }
}

fn impure_builtin_label(node: &AstNode<'_>, ctx: &LintContext<'_>) -> Option<String> {
    match node.kind() {
        AstKind::NewExpression(new_expression) => {
            let Expression::Identifier(callee) = &new_expression.callee else {
                return None;
            };
            (callee.name == "Date"
                && new_expression.arguments.is_empty()
                && is_global_reference(callee, ctx))
            .then(|| "new Date()".to_string())
        }
        AstKind::CallExpression(call_expression) => {
            if let Expression::Identifier(callee) = &call_expression.callee
                && callee.name == "Date"
                && call_expression.arguments.is_empty()
                && is_global_reference(callee, ctx)
            {
                return Some("Date()".to_string());
            }
            let member_expression = call_expression.callee.as_member_expression()?;
            let Expression::Identifier(receiver) = member_expression.object().get_inner_expression()
            else {
                return None;
            };
            let property_name = member_expression.static_property_name()?;
            (IMPURE_MEMBER_CALLS
                .iter()
                .any(|(namespace, method)| receiver.name == *namespace && property_name == *method)
                && is_global_reference(receiver, ctx))
            .then(|| format!("{}.{}()", receiver.name, property_name))
        }
        _ => None,
    }
}

fn is_global_reference(
    identifier: &oxc_ast::ast::IdentifierReference<'_>,
    ctx: &LintContext<'_>,
) -> bool {
    ctx.scoping()
        .get_reference(identifier.reference_id())
        .symbol_id()
        .is_none()
}

fn is_module_scope_evaluation(node: &AstNode<'_>, ctx: &LintContext<'_>) -> bool {
    let mut child = node;
    let mut cursor = ctx.nodes().parent_node(node.id());
    loop {
        match cursor.kind() {
            AstKind::Function(_) | AstKind::ArrowFunctionExpression(_) => {
                let function_root = transparent_expression_root(cursor, ctx);
                let call_node = ctx.nodes().parent_node(function_root.id());
                if !matches!(
                    call_node.kind(),
                    AstKind::CallExpression(call_expression)
                        if call_expression.callee.span() == function_root.span()
                ) {
                    return false;
                }
                child = call_node;
                cursor = ctx.nodes().parent_node(call_node.id());
                continue;
            }
            AstKind::MethodDefinition(_) => return false,
            AstKind::ConditionalExpression(conditional) if conditional.test.span() != child.span() => {
                let server_test_value = server_value_of_browser_global_test(
                    &conditional.test,
                    ctx,
                    &mut Vec::new(),
                );
                if server_test_value.is_some_and(|server_value| {
                    conditional.consequent.span() == child.span() && !server_value
                        || conditional.alternate.span() == child.span() && server_value
                }) {
                    return false;
                }
            }
            AstKind::LogicalExpression(logical)
                if logical.right.span() == child.span()
                    && matches!(logical.operator, LogicalOperator::And | LogicalOperator::Or) =>
            {
                let server_test_value =
                    server_value_of_browser_global_test(&logical.left, ctx, &mut Vec::new());
                if server_test_value.is_some_and(|server_value| {
                    logical.operator == LogicalOperator::And && !server_value
                        || logical.operator == LogicalOperator::Or && server_value
                }) {
                    return false;
                }
            }
            AstKind::PropertyDefinition(property) => {
                if !property.r#static
                    || property
                        .value
                        .as_ref()
                        .is_none_or(|value| !value.span().contains_inclusive(child.span()))
                {
                    return false;
                }
            }
            AstKind::VariableDeclarator(declarator) => {
                let binding = declarator.id.get_binding_identifier();
                let is_top_level_binding = binding.is_some_and(|binding| {
                    ctx.scoping()
                        .scope_flags(ctx.scoping().symbol_scope_id(binding.symbol_id()))
                        .is_top()
                });
                let declaration = ctx.nodes().parent_node(cursor.id());
                let AstKind::VariableDeclaration(variable_declaration) = declaration.kind() else {
                    return false;
                };
                if !variable_declaration.kind.is_const() {
                    let Some(binding) = binding else {
                        return false;
                    };
                    if ctx
                        .scoping()
                        .get_resolved_references(binding.symbol_id())
                        .any(oxc_semantic::Reference::is_write)
                    {
                        return false;
                    }
                }
                if is_top_level_binding {
                    return true;
                }
                let mut declaration_parent = ctx.nodes().parent_node(declaration.id());
                if matches!(declaration_parent.kind(), AstKind::ExportNamedDeclaration(_)) {
                    declaration_parent = ctx.nodes().parent_node(declaration_parent.id());
                }
                return matches!(declaration_parent.kind(), AstKind::Program(_));
            }
            AstKind::Program(_) => return true,
            _ => {}
        }
        child = cursor;
        cursor = ctx.nodes().parent_node(cursor.id());
    }
}

fn server_value_of_browser_global_test<'a>(
    expression: &Expression<'a>,
    ctx: &LintContext<'a>,
    visited_symbol_ids: &mut Vec<oxc_semantic::SymbolId>,
) -> Option<bool> {
    match expression.get_inner_expression() {
        Expression::Identifier(identifier) => {
            let symbol_id = ctx
                .scoping()
                .get_reference(identifier.reference_id())
                .symbol_id()?;
            if visited_symbol_ids.contains(&symbol_id) {
                return None;
            }
            let initializer = resolve_direct_unreassigned_symbol_initializer(symbol_id, ctx)?;
            visited_symbol_ids.push(symbol_id);
            let result = server_value_of_browser_global_test(initializer, ctx, visited_symbol_ids);
            visited_symbol_ids.pop();
            result
        }
        Expression::UnaryExpression(unary) if unary.operator == UnaryOperator::LogicalNot => {
            server_value_of_browser_global_test(&unary.argument, ctx, visited_symbol_ids)
                .map(|value| !value)
        }
        Expression::LogicalExpression(logical) => {
            let left = server_value_of_browser_global_test(
                &logical.left,
                ctx,
                &mut visited_symbol_ids.clone(),
            )?;
            match logical.operator {
                LogicalOperator::And if !left => Some(false),
                LogicalOperator::Or if left => Some(true),
                LogicalOperator::And | LogicalOperator::Or => server_value_of_browser_global_test(
                    &logical.right,
                    ctx,
                    visited_symbol_ids,
                ),
                _ => None,
            }
        }
        Expression::BinaryExpression(binary) => {
            if !is_undefined_typeof_browser_global(&binary.left, ctx)
                && !is_undefined_typeof_browser_global(&binary.right, ctx)
            {
                return None;
            }
            let has_undefined_literal = matches!(
                binary.left.get_inner_expression(),
                Expression::StringLiteral(literal) if literal.value == "undefined"
            ) || matches!(
                binary.right.get_inner_expression(),
                Expression::StringLiteral(literal) if literal.value == "undefined"
            );
            if !has_undefined_literal {
                return None;
            }
            match binary.operator {
                BinaryOperator::Equality | BinaryOperator::StrictEquality => Some(true),
                BinaryOperator::Inequality | BinaryOperator::StrictInequality => Some(false),
                _ => None,
            }
        }
        _ => None,
    }
}

fn is_undefined_typeof_browser_global(
    expression: &Expression<'_>,
    ctx: &LintContext<'_>,
) -> bool {
    let Expression::UnaryExpression(unary) = expression.get_inner_expression() else {
        return false;
    };
    let Expression::Identifier(identifier) = unary.argument.get_inner_expression() else {
        return false;
    };
    unary.operator == UnaryOperator::Typeof
        && matches!(identifier.name.as_str(), "window" | "document")
        && is_global_reference(identifier, ctx)
}
